File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
from collections import deque

def solution(N, number):
    answer = -1

    dp = []
    
    for cnt in range(1, 9):
        tmp = [int(str(N) * cnt)]
        for i in range(1, cnt):
            left, right = dp[i-1], dp[cnt - i - 1]
            
            for l in left:
                for r in right:
                    case = [l + r, l - r, l * r, l // r]
                    for c in case:
                        if c > 0:
                            tmp.append(c)
        dp.append(tmp)
        if number in tmp:
            return cnt
    

    return answer

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int N, int number) {
        int answer = 0;
        
        Set<Integer> dp[] = new Set[9];
        for (int i = 0; i < 9; i++) dp[i] = new HashSet();
        
        for (int i = 1; i < 9; i++) {
            
            String num = "";
            
            for (int j = 0; j < i; j++) num += Integer.toString(N);
            
            dp[i].add(Integer.parseInt(num));
        }
        if (N == number) return 1;
        
        for (int i = 1; i < 9; i++) {
            String num = "";    
            for (int j = 0; j < i; j++) num += Integer.toString(N);
            
            if (Integer.parseInt(num) == number) return i;
            for (int left = 1; left < i; left++) {
                Set<Integer> leftSet = dp[left];
                Set<Integer> rightSet = dp[i - left];
                
                for (int leftNum : leftSet) {
                    for (int rightNum: rightSet) {
                        int[] tmp = new int[] {
                            leftNum + rightNum,
                            leftNum - rightNum,
                            leftNum * rightNum,
                            leftNum / rightNum
                        };
                        
                        for (int t : tmp) {
                            if (t == number) return i;
                            if (t == 0) continue;
                            dp[i].add(t);
                        }
                    }
                }
            }
        }
        
        return -1;
    }
}